### PR TITLE
fix: fix some linearization bugs

### LINKED ISSUE #2907: undefvar and bounds error in linearization_function
https://github.com/SciML/ModelingToolkit.jl/blob/a7f6cd2b09fe1c36ce86f05d81b24004ee5fd961/src/systems/abstractsystem.jl#L2170

@AayushSabharwal  isn't there something missing in the expression `let p_getter,`?, e.g., `let p_getter = p_getter,`?

Making that change, I further get 
```julia
ERROR: BoundsError: attempt to access Tuple{SymbolicIndexingInterface.ProblemState{Vector{Float64}, ModelingToolkit.MTKParameters{Tuple{Vector{Float64}}, Tuple{}, Tuple{}, Tuple{}, Tuple{}, RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:ˍ₋out, :ˍ₋arg1), ModelingToolkit.var"#_RGF_ModTag", ModelingToolkit.var"#_RGF_ModTag", (0x436e531f, 0xa1112b08, 0x3bf49a23, 0x855dd196, 0x09507fa9), Nothing}, RuntimeGeneratedFunctions.RuntimeGeneratedFunction{(:ˍ₋arg1,), ModelingToolkit.var"#_RGF_ModTag", ModelingToolkit.var"#_RGF_ModTag", (0xee8ca58d, 0x1958d6ce, 0x314ac4d0, 0xd540ea53, 0x6f246bd0), Nothing}}, Float64}} at index [2]
Stacktrace:
  [1] getindex(t::Tuple, i::Int64)
    @ Base ./tuple.jl:31
  [2] macro expansion
    @ ~/.julia/packages/RuntimeGeneratedFunctions/M9ZX8/src/RuntimeGeneratedFunctions.jl:162 [inlined]
  [3] macro expansion
    @ ./none:0 [inlined]
  [4] generated_callfunc
    @ ./none:0 [inlined]
  [5] (::RuntimeGeneratedFunctions.RuntimeGeneratedFunction{…})(args::SymbolicIndexingInterface.ProblemState{…})
    @ RuntimeGeneratedFunctions ~/.julia/packages/RuntimeGeneratedFunctions/M9ZX8/src/RuntimeGeneratedFunctions.jl:150
  [6] (::ModelingToolkit.var"#1253#1261"{…})(u::Vector{…}, p::ModelingToolkit.MTKParameters{…}, t::Float64)
    @ ModelingToolkit ~/.julia/dev/ModelingToolkit/src/systems/abstractsystem.jl:2176
  [7] (::ModelingToolkit.var"#1255#1263"{…})(u::Vector{…}, p::Dict{…}, t::Float64)
    @ ModelingToolkit ~/.julia/dev/ModelingToolkit/src/systems/abstractsystem.jl:2231
  [8] linearize(sys::ODESystem, lin_fun::ModelingToolkit.var"#1255#1263"{…}; t::Float64, op::Dict{…}, allow_input_derivatives::Bool, p::SciMLBase.NullParameters)
    @ ModelingToolkit ~/.julia/packages/ModelingToolkit/3AmNp/src/systems/abstractsystem.jl:2385
```
reproducer exists, but it's not very minimal

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Close #2907 

## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

File: src/systems/abstractsystem.jl [hunk 1]
```diff
@@ -2167,14 +2167,14 @@ function linearization_function(sys::AbstractSystem, inputs,
         u_getter = isempty(unknowns(initsys)) ? (_...) -> nothing :
                    build_explicit_observed_function(
             sys, unknowns(initsys); eval_expression, eval_module)
-        get_initprob_u_p = let p_getter,
+        get_initprob_u_p = let p_getter = p_getter,
             p_setter! = setp(initsys, initsys_ps),
             u_getter = u_getter
 
             function (u, p, t)
                 state = ProblemState(; u, p, t)
```
Comment (baggepinnen) [inline src/systems/abstractsystem.jl:2175]: the `state` is now defined but not used?

Comment (AayushSabharwal) [reply src/systems/abstractsystem.jl]: Thanks for pointing that out

**Comment by ChrisRackauckas** [comment]
Isn't this also supposed to handle the SciMLBase downstream https://github.com/SciML/ModelingToolkit.jl/actions/runs/10177059599/job/28147675146?pr=2912#step:6:2107 ?

**Comment by AayushSabharwal** [comment]
No, that's in SII and is merged. Was waiting for CI on https://github.com/SciML/SymbolicIndexingInterface.jl/pull/90 to be able to tag that too

**Comment by AayushSabharwal** [comment]
https://github.com/JuliaRegistries/General/pull/112093

**Pull request merged**